pil-analyzer/src/evaluator.rs

```diff
@@ -1733,4 +1760,131 @@ mod test {
+
+            let x: int -> fe = |q| match Add::add(q, 2) {
+                     v => std::convert::fe(v),
+            };
+
+            let y: int -> fe = |q| match Add::add(q, 4) {
+                    v => x(v) + Cast::add(v),
+            };
+
+            let r: fe = y(2);
+        ";
+
+        assert_eq!(parse_and_evaluate_symbol(input, "F::r"), "14".to_string());
+    }
+
+    #[test]
+    fn traits_multiple_fns() {
+        let input = "
+        namespace std::convert(4);
+            let fe = || fe();
+        namespace F(4);
+            trait Do<T, Q> {
+                add: T, T -> Q,
+                sub: T, T -> Q,
+                cast: T -> Q,
+            }
+
+            impl Do<int, fe> {
+                add: |a, b| std::convert::fe(a + b),
+                sub: |a, b| std::convert::fe(a - b),
+                cast: |a| std::convert::fe(a),
+            }
+
+            let x: int -> fe = |q| match Do::sub(q, q) {
+                v => {
+                    let one: int = 1;
+                    let two: int = 2 * 1;
+                    v + Do::add(one, two)
+                },
+            };
+
+            let y: int -> fe = |q| match Do::cast(q) {
+                v => { 
+                    let two: int = 2;
+                    x(two) + v
+                },
+            };
+
+            let r: fe = y(2);
+        ";
+
+        assert_eq!(parse_and_evaluate_symbol(input, "F::r"), "5".to_string());
+    }
+
+    #[test]
+    fn traits_multiple_impls() {
+        let input = "
+        namespace std::convert(4);
+            let fe = || fe();
+        namespace F(4);
+            trait Do<T, Q> {
+                add: T, T -> Q,
+                sub: T, T -> Q,
+                cast: T -> Q,
+            }
+
+            impl Do<int, fe> {
+                add: |a, b| std::convert::fe(a + b),
+                sub: |a, b| std::convert::fe(a - b),
+                cast: |a| std::convert::fe(a),
+            }
+
+            impl Do<int, int> {
+                add: |a, b| a + b + 1,
+                sub: |a, b| a - b + 1,
+            }
+
+            impl Do<fe, fe> {
+                add: |a, b| a + b + std::convert::fe(6),
+                sub: |a, b| a - b - std::convert::fe(1),
+            }
+
+            let x: int -> fe = |q| match Do::sub::<int, int>(q, q) {
+                v => {
+                    let p1: fe = 3;
+                    let p2: fe = 2 * 1;
+                    Do::add(p1, p2)
+                },
+            };
+
+            let y: int -> int = |q| match q {
+                v => {
+                    let p1: int = 5 - v;
+                    let p2: int = 2 * 1;
+                    Do::add(p1, p2)
+                },
+            };
+
+            let z: int = y(2);
+            let r: fe = x(2);
+            ";
+
+        assert_eq!(parse_and_evaluate_symbol(input, "F::r"), "11".to_string());
+        assert_eq!(parse_and_evaluate_symbol(input, "F::z"), "6".to_string());
+    }
 }
```

Comment:
Please add a test where you call a trait function inside a trait impl, because I think there is still something missing in the pil analyzer regarding that.